Source mapping issues for large / deeply-nested templates

With
```
    "@glint/core": "2.0.0-alpha.6",
    "@glint/template": "1.6.0-alpha.2",
```
if I create a `foo.gts` file with the follow contents:
```hbs
<template>
  <div>
    <div>
      <div>
        <div>
          <div>
            <div>
              <ul>
                <li>Use the Office action as a
                  <strong>Word (docx) file</strong>
                  that was downloaded from Patent Center (typically the best and
                  fastest option as it doesn't require OCR)</li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>
```
`glint` produces the following error:
```
app/components/foo.gts:12:53 - error TS2345: Argument of type '{}' is not assignable to parameter of type 'never'.

12                   fastest option as it doesn't require OCR)</li>
                                                       ~~~~~


Found 1 error in app/components/foo.gts:12
```
This a minimal repro for a condition I found in my application. Almost any tweak that I make to the file (removing a `<div>` level, deleting or adding a character anywhere) seems to "fix" the error, but this exact content reproduces it. Very bizarre.

This does not reproduce with the release version of `@glint/*`:
```
    "@glint/core": "^1.2.1",
    "@glint/environment-ember-loose": "^1.2.1",
    "@glint/environment-ember-template-imports": "^1.3.0",
    "@glint/template": "^1.2.1",
```

--- a/packages/core/src/transform/template/transformed-module.ts
+++ b/packages/core/src/transform/template/transformed-module.ts
@@ -343,26 +343,33 @@ export default class TransformedModule {
       // of generated code map back to the same source region.
       const hbsLength = hbsEnd - hbsStart;
       const tsLength = tsEnd - tsStart;
-      if (hbsLength === tsLength) {
-        // (Hacky?) assumption: because TS and HBS span lengths are equivalent,
-        // then this is a simple leafmost mapping, e.g. `{{this.[foo]}}` -> `this.[foo]`
-        push(hbsStart, tsStart, hbsLength, mapping.codeInformation);
-      } else {
-        // Disregard the "null zone" mappings, i.e. cases where TS code maps to empty HBS code
-        if (hbsLength > 0 && tsLength > 0) {
-          push(hbsStart, tsStart, 0, mapping.codeInformation);
-          push(hbsEnd, tsEnd, 0, mapping.codeInformation);
-        }
-      }
+
+      // Disregard the "null zone" mappings, i.e. cases where TS code maps to empty HBS code
+      const isNonNullMapping = hbsLength > 0 && tsLength > 0;
 
       if (children.length > 0) {
-        push(hbsStart, tsStart, 0, mapping.codeInformation);
+        if (isNonNullMapping) {
+          push(hbsStart, tsStart, 0, mapping.codeInformation);
+        }
 
         mapping.children.forEach((child) => {
           recurse(span, child);
         });
 
-        push(hbsEnd, tsEnd, 0, mapping.codeInformation);
+        // Disregard the "null zone" mappings, i.e. cases where TS code maps to empty HBS code
+        if (isNonNullMapping) {
+          push(hbsEnd, tsEnd, 0, mapping.codeInformation);
+        }
+      } else {
+        if (hbsLength === tsLength) {
+          push(hbsStart, tsStart, hbsLength, mapping.codeInformation);
+        } else {
+          // Disregard the "null zone" mappings, i.e. cases where TS code maps to empty HBS code
+          if (isNonNullMapping) {
+            push(hbsStart, tsStart, 0, mapping.codeInformation);
+            push(hbsEnd, tsEnd, 0, mapping.codeInformation);
+          }
+        }
       }
     };
 

